fix(ngcc): dts declaration mapping

petebacondarwin opened the pull request:
The naïve matching algorithm we previously used to match declarations in
source files to declarations in typings files was based only on the name
of the thing being declared.  This did not handle cases where the declared
item has been exported via an alias. This is a common scenariow when one
of the two file sets (source or typings) has been flattened, while the other
has not.

The new algorithm tries to overcome this by creating two maps of export
name to declaration (i.e. `Map<string, ts.Declaration>`).
One for the source files and one for the typings files.
It then joins these two together by matching export names, resulting in a
new map that maps source declarations to typings declarations directly
(i.e. `Map<ts.Declaration, ts.Declaration>`).

This new map can handle the declaration names being different between the
source and typings as long as they are ultimately both exported with the
same alias name.

Fixes #33593
Fixes #34191
Fixes #32442

petebacondarwin commented on packages/compiler/src/render3/r3_pipe_compiler.ts:
```diff
@@ -64,14 +64,14 @@ export function compilePipeFromMetadata(metadata: R3PipeMetadata) {
 
   // e.g. `type: MyPipe`
-  definitionMapValues.push({key: 'type', value: metadata.type, quoted: false});
+  definitionMapValues.push({key: 'type', value: metadata.type.value, quoted: false});
```
I think this should be `type.value` but I am not 100% sure. What do you think @alxhub ?

alxhub replied:
You're correct. :)

gkalpak commented on packages/compiler-cli/ngcc/test/rendering/dts_renderer_spec.ts:
```diff
@@ -114,7 +114,7 @@ runInEachFileSystem(() => {
       };
       INPUT_DTS_PROGRAM = {
-        name: _('/typings/file.d.ts'),
+        name: _('/node_modules/test-package/typings/file.d.ts'),
```
OOC, why is this change needed?

petebacondarwin replied:
Because of this line: https://github.com/angular/angular/pull/34254/commits/cf36bfc0d982dbfc06f4dab782e6f3ccda59342f#diff-56411fd1f865093a818fbf15a5ea15f6R1998

```


/**
 * Execute the `callBack` for each file in the bundle program, starting with the root file.
 *
 * @param bundle the bundle program whose files should be iterated over
 * @param callBack the function to all for each of the files in the program
 */
function forEachFileInProgram(bundle: BundleProgram, callBack: (file: ts.SourceFile) => void) {
  const rootFile = bundle.program.getSourceFile(bundle.path);
  if (!rootFile) {
    throw new Error(`The given rootPath ${rootFile} is not a file of the given program.`);
  }
  // First run the call back on the root file.
  callBack(rootFile);
  bundle.program.getSourceFiles().forEach(file => {
    // Ignore if the file is the root (already processed) or if the file is not part of the package.
    if (file !== rootFile && isWithinPackage(bundle.package, file)) {
      callBack(file);
    }
  });
}
```

When computing the dts mappings we now only consider files that are "within" the package.

gkalpak reviewed (approved):
LGTM for `fw-ngcc` 👍

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -83,9 +83,9 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
-      dts?: BundleProgram|null) {
-    super(checker);
+      protected logger: Logger, protected isCore: boolean, src: BundleProgram,
+      dts: BundleProgram|null = null) {
```
Before this the type signature was `BundleProgram|null|undefined` because of the `dts?:`.
There was no tangible difference between `null` and `undefined` here so I dropped the `undefined` by this change.

gkalpak replied:
I see. I don't think it makes any different in the commit it was introduced in, but it does make sense in tandem with changes in a subsequent commit where `dts` is made a protected property of the class and the [this.dtsDeclarationMap = ...][1] line below is removed.

<sub>(I am fine keeping it here for the sake of reducing commit noise.)</sub>

[1]: https://github.com/angular/angular/pull/34254/commits/6c44955b4#diff-56411fd1f865093a818fbf15a5ea15f6R89

petebacondarwin replied:
Oh sorry - a rebase mixup. Thanks for being generous :-)

petebacondarwin replied:
I moved it to the other commit (and updated the other hosts too) since I needed to do another rebase anyway. (No fixup commits, sorry)

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/transformer.ts:
```diff
@@ -103,15 +103,13 @@ export class Transformer {
     const typeChecker = bundle.src.program.getTypeChecker();
```
Thanks!
<sub>I assume you meant "not used"</sub>

petebacondarwin commented on packages/compiler-cli/ngcc/src/analysis/private_declarations_analyzer.ts:
```diff
@@ -58,33 +58,11 @@ export class PrivateDeclarationsAnalyzer {
                 // This declaration is public so we can remove it from the list
                 privateDeclarations.delete(declaration.node.name);
-              } else if (!this.host.getDtsDeclaration(declaration.node)) {
+              } else {
```
Actually I think this entire `else` block (and associated `alias` property on `ExportInfo` can now be dropped since the dts mapping is able to handle aliasing.

petebacondarwin commented:
Rebased on top of #34356 as that is also needed to fix #33593

alxhub commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1972,3 +1979,24 @@ function getContainingStatement(node: ts.Node): ts.ExpressionStatement|null {
+ * @param bundle the bundle program whose files should be iterated over
+ * @param callBack the function to all for each of the files in the program
+ */
+function forEachFileInProgram(bundle: BundleProgram, callBack: (file: ts.SourceFile) => void) {
```
```suggestion
function forEachFileInProgram(bundle: BundleProgram, callback: (file: ts.SourceFile) => void) {
```

alxhub commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1972,3 +1979,24 @@ function getContainingStatement(node: ts.Node): ts.ExpressionStatement|null {
+  }
+  // First run the call back on the root file.
+  callBack(rootFile);
+  bundle.program.getSourceFiles().forEach(file => {
```
Should we check `isDeclarationFile` too?

petebacondarwin replied:
No because this function is called twice - once with dts files and once with source files.

alxhub commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1499,42 +1507,25 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+      Map<ts.Declaration, ts.Declaration> {
+    const dtsMap = computeDtsExportDeclarationMap(dts);
+    const declarationMap = new Map<ts.Declaration, ts.Declaration>();
+    forEachFileInProgram(src, file => {
```
Is this logic correct? Only the entrypoint index's exports are expected to match up to the .d.ts program, I would think.

So if the source program has a file `lib/internal/foo.js` that has `export class Foo {...}`, there is zero guarantee that this is the same `Foo` exported from the root of the `.d.ts` program.

I see that `forEachFileInProgram` will process the root file first. For that file, the logic will be correct.

For other files, though, we're making an assumption that export names will line up. That's okay, but we should be clear about such assumptions.

I think we should:

1) keep the `Map`s of certified matches vs guesses separate. So we should have a primary `Map` that matches up the root file, and a "guess" `Map` that matches up deeply exported things by name only.

2) currently, the logic actually overwrites `declarationMap` unilaterally on a match, even if previously the root file gave us a definitive (and different) answer. The split approach above will help avoid this problem.

3) we should have a comment here explaining the matching algorithm and the two different approaches.

petebacondarwin replied:
I have refactored this as suggested. There are now two maps: one for exports that are exposed publicly via the entry-point (i.e. the root file) and one for exports that are exposed from individual files in the source tree. 

As a result of doing this I found that there were further problems with the CommonJS and UMD re-export handling, which I have fixed in two extra commits.

Therefore the whole of this PR could probably do with another review: @alxhub , @gkalpak , @JoostK .

alxhub commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1863,21 +1854,45 @@ function isClassMemberType(node: ts.Declaration): node is ts.ClassElement|
+ * @param dtsProgram The program containing all the dts declarations.
+ * @returns a map of exported names to declarations.
+ */
+function computeDtsExportDeclarationMap(dts: BundleProgram): Map<string, ts.Declaration> {
```
I think the same goes for .d.ts exports. We should keep entrypoint exports and deep exports in separate `Map`s.

alxhub reviewed (changes requested):
Sorry, wrong button.

petebacondarwin committed:
refactor(ngcc): use bundle src to create reflection hosts

Previously individual properties of the src bundle program were
passed to the reflection host constructors. But going forward,
more properties will be required. To prevent the signature getting
continually larger and more unwieldy, this change just passes the
whole src bundle to the constructor, allowing it to extract what it
needs.

petebacondarwin committed:
fix(ngcc): handle CommonJS re-exports by reference

In TS we can re-export imports using statements of the form:

```
export * from 'some-import';
```

This can be downleveled in CommonJS to either:

```
__export(require('some-import'));
```

or

```
var someImport = require('some-import');
__export(someImport);
```

Previously we only supported the first downleveled version.
This commit adds support for the second version.

petebacondarwin committed:
fix(ngcc): handle UMD re-exports

In TS we can re-export imports using statements of the form:

```
export * from 'some-import';
```

This is downleveled in UMD to:

```
function factory(exports, someImport) {
  function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
  }
  __export(someImport);
}
```

This commit adds support for this.

petebacondarwin committed:
fix(ngcc): correctly match aliased classes between src and dts files

The naïve matching algorithm we previously used to match declarations in
source files to declarations in typings files was based only on the name
of the thing being declared.  This did not handle cases where the declared
item had been exported via an alias - a common scenario when one of the two
file sets (source or typings) has been flattened, while the other has not.

The new algorithm tries to overcome this by creating two maps of export
name to declaration (i.e. `Map<string, ts.Declaration>`).
One for the source files and one for the typings files.
It then joins these two together by matching export names, resulting in a
new map that maps source declarations to typings declarations directly
(i.e. `Map<ts.Declaration, ts.Declaration>`).

This new map can handle the declaration names being different between the
source and typings as long as they are ultimately both exported with the
same alias name.

Further more, there is one map for "public exports", i.e. exported via the
root of the source tree (the entry-point), and another map for "private
exports", which are exported from individual files in the source tree but
not necessarily from the root. This second map can be used to "guess"
the mapping between exports in a deep (non-flat) file tree, which can be
used by ngcc to add required private exports to the entry-point.

Fixes #33593

petebacondarwin committed:
fix(ngcc): use the correct identifiers when updating typings files

Previously the identifiers used in the typings files were the same as
those used in the source files.

When the typings files and the source files do not match exactly, e.g.
when one of them is flattened, while the other is a deep tree, it is
possible for identifiers to be renamed.

This commit ensures that the correct identifier is used in typings files
when the typings file does not export the same name as the source file.

Fixes https://github.com/angular/ngcc-validation/pull/608

petebacondarwin committed:
refactor(ngcc): slightly improve the info in error messages

petebacondarwin committed:
refactor(ngcc): remove private declaration aliases

Now that the source to typings matching is able to handle
aliasing of exports, there is no need to handle aliases in private
declarations analysis.

These were originally added to cope when the typings files had
to use the name that the original source files used when exporting.

alxhub reviewed (approved):
👍 I really like the logic here - it's clean, easy to follow, and I think will address a lot of subtle .d.ts matching issues we have today.

alxhub commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:286221348:BASE:286221394:1576694400309:aa213af6)
[Ivy Presubmit](https://test.corp.google.com/ui#id=OCL:286221348:BASE:286221394:1576694400309:c0c11529)

gkalpak commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -237,6 +273,15 @@ interface UmdExportDeclaration {
+      statement.expression.expression.text === '__export' &&
+      statement.expression.arguments.length === 1 &&
+      ts.isIdentifier(statement.expression.arguments[0]);
+}
```
Can't UMD have `__export(require('...'))` type re-exports? I thought it could.

petebacondarwin replied:
You can have commonjs `require` imports but they must be passed to the factory function. I dont see how you could have `require` reexports without breaking non-commonjs formats

gkalpak replied:
My understanding is that you could also have `require(...)` calls inside the factory (what RequireJS calls [simplified CommonJS wrapping][1]). Also, the TypeScript playground seems to generate that format for UMD (but that might be because the playground doesn't do a proper build): [TS playground example][2]

[1]: https://requirejs.org/docs/whyamd.html#sugar 
[2]: https://www.typescriptlang.org/play/?target=1&module=3&ssl=1&ssc=1&pln=28&pc=1#code/JYWwDg9gTgLgBAIhgOwM4FoDG0Cm6QQAmArgDY6oD0AhmGKcJtTMBMggNwBQXOAHpFhwAVHABmUCCEQA6SvWZjoILFILtu-QfFESps+aUXLOvAdB3jJ0hHIUwlUFahjNGprRZFX9tw8ad0MEkAN2BCHChUD3MhXWsDKAoIYihMPFIIagioUzNtbz0bOWQ3EIpMKGAwGBkCEnIYgvjfOWwQdTzPOJ9iykliGEimrxa+sWBydFQATxccEBGeooMACxgYMCwGHGQYJcsVv0cQaO58rwBvLjhbuAA1YBwAdwBhI1RUABobu8eX97UT4AWRwrh+dweT2eACVkqRylAIXcAOZg-5vD7fX63NEwDGg8E4uDAVAAaWQEGeyAxyNuSRRpKGUAAouQQLsYD8AL69Aw4dmc9AMplQGZ5bBoeBKCBwAC8cAA2jIVQByACMACYAMyqgC63CAA

petebacondarwin replied:
OK, so AMD supports a `require` argument passed through to the factory function, so 
I guess it is possible and we should perhaps support it... But in that case you cannot use the UMD templates that support globals. See https://github.com/umdjs/umd

gkalpak replied:
I am not arguing whether it is a good idea or not to use `require()` calls in UMD, but if it is possible that the UMD files we need to process use that, then it would be better if we detected it.

I'd be happy to add support for this in a PR I am working on to fix star re-exports with imported helpers.

gkalpak replied:
For reference: #34512
